### PR TITLE
Check for duplicate UACs

### PR DESCRIPTION
**Pull request by AdamHawtin**
# Motivation and Context
There are circumstances where action rule messages can get ingested twice, causing duplicate rows on the partial print files, which could then meet its expected row count erroneously. We need to log an error in this instance and quarantine the file.

# What has changed
* Check for duplicate UAC's in the partial print file before continuing with it
* Quarantine any files with duplicate UAC's and log an error

# How to test?
Try manually publishing action rule messages with duplicate UACs in the same batch, you should see the resulting print file gets quarantined instead of being encrypted and sent.

# Links
https://trello.com/c/3OuAQeMC/973-investigate-duplicate-rows-in-print-file-service-5

### CODE REVIEW

**Comment by AdamHawtin** [comment]
[Investigate duplicate rows in print file service (5)](https://trello.com/c/3OuAQeMC/973-investigate-duplicate-rows-in-print-file-service-5)

File: test/integration_tests/test_end_to_end.py [hunk 1]
```diff
@@ -252,7 +246,58 @@ def test_our_decryption_key():
         decryption_key_path=Path(__file__).parents[2].joinpath('dummy_keys',
                                                                'our_dummy_private.asc'),
         decryption_key_passphrase='test',
-        expected='test_uac|test_caseref|Mr|Test|McTest|123 Fake Street|Duffryn||Newport|NPXXXX|P_IC_ICL1\n')
+        expected='0|test_caseref|Mr|Test|McTest|123 Fake Street|Duffryn||Newport|NPXXXX|P_IC_ICL1\n')
+
+
+def test_icl1e_with_duplicate_uacs_is_quarantined():
```
Comment (jcox-dev) [inline test/integration_tests/test_end_to_end.py]: Might be worth adding a sanity check in one / both of these to make sure a file hasn't been created elsewhere other than the quarantine.

File: app/file_sender.py [hunk 2]
```diff
@@ -82,25 +92,18 @@ def copy_files_to_sftp(file_paths: Collection[Path], remote_directory):
...
-        ]
-    }
+def check_partial_has_no_duplicates(partial_file_path: Path, pack_code: str):
+    uacs = set()
+    fieldnames = DATASET_TO_PRINT_TEMPLATE[PACK_CODE_TO_DATASET[pack_code]]
+    with open(partial_file_path) as partial_file:
+        reader = csv.DictReader(partial_file, fieldnames=fieldnames, delimiter='|')
+        uac_columns = {fieldname for fieldname in fieldnames if 'uac' in fieldname}
+        for line_number, row in enumerate(reader, 1):
+            if any((row.get(uac_column) and row.get(uac_column) in uacs) for uac_column in uac_columns):
```
Comment (jcox-dev) [inline app/file_sender.py]: Nuking `None`s. Nice.

File: docker-compose.yml [hunk 3]
```diff
@@ -34,9 +34,11 @@ services:
       - SFTP_KEY_FILENAME=/app/dummy_keys/dummy_rsa
       - SFTP_PASSPHRASE=secret
       - ENVIRONMENT=DEV
-    external_links:
-      - dev-print-file-service-rabbitmq
-      - dev-print-file-service-sftp
+      - QUARANTINED_FILES_DIRECTORY=/working_files/quarantined_files
```
Comment (jcox-dev) [inline docker-compose.yml]: 👍

File: test/integration_tests/test_end_to_end.py (hunk 1 above)
Comment (AdamHawtin) [reply test/integration_tests/test_end_to_end.py]: I was wondering about that, but I was unsure how long I should leave the test polling for files before accepting that as evidence they'll never be there.

Comment (jcox-dev) [reply test/integration_tests/test_end_to_end.py]: I was more suggesting that we should be checking that files are never in two places but perhaps that's over zealous.

**Pull request merged**